File: tests/orig_parity/11_nhefs_evalue.py
"""StatsPAI original-data parity (Python side) -- Module 11.

E-value sensitivity analysis (VanderWeele & Ding, *Ann Intern Med* 2017)
demonstrated with ``sp.evalue`` on the REAL NHEFS data bundled in
``sp.datasets.nhefs()``.

The E-value answers: *how strong would an unmeasured confounder have to
be* -- on the risk-ratio scale, with both treatment and outcome -- *to
fully explain away an observed association?*  For a risk ratio RR >= 1 it
has the deterministic closed form

    E = RR + sqrt(RR * (RR - 1))                                   (VD-2017)

(and the same formula applied to 1/RR when RR < 1).  Because it is a
fixed function of the input ratio, ``sp.evalue`` must reproduce the
``EValue`` R package and that closed form to ~1e-10 -- this module pins
that.

PRIMARY (clean ratio measure):  the IP-weighted 10-year mortality risk
ratio of quitting smoking (``qsmk``) vs not, on the FULL NHEFS
(n=1629).  Stabilized IP weights from the canonical book confounder
model give RR_adj ~ 0.99 (death risk is essentially unchanged after
adjustment -- the well-known "null after confounding control" result);
its CI straddles 1, so the E-value collapses to 1.00.  We therefore ALSO
report the *crude* mortality RR (~1.33, before adjustment, CI excludes
1), whose E-value (~1.98) is the pedagogically interesting number and
the one whose closed form we anchor.

SECONDARY (continuous outcome):  the Chapter-12 weight effect (book 3.4
kg; our IPW point 3.44 kg).  ``sp.evalue`` natively accepts a continuous
effect via ``measure='SMD'``, implementing the VanderWeele-Ding
standardized-mean-difference approximation ``RR_approx = exp(0.91 * d)``
with ``d = effect / SD(outcome)``.  We pass the standardized effect and
confirm it equals both the manual ``exp(0.91 d)`` route and
``EValue::evalues.MD``.

Anchors per statistic: StatsPAI (sp.evalue) | R gold (EValue package +
hand-rolled stabilized IPW) on the same CSV bytes | the deterministic
closed form / published convention.
"""

from __future__ import annotations

import math

import numpy as np
import statsmodels.api as sm
import statspai as sp

from _common import OrigRecord, write_results
from _nhefs import book_design

MODULE = "11_nhefs_evalue"

# VanderWeele-Ding standardized-mean-difference -> approximate risk ratio.
VD_SMD_COEF = 0.91


def _stabilized_ipw_rr(dd, covs, y: str, treat: str) -> tuple[float, float, float]:
    """Hand-rolled stabilized-IPW risk ratio (book Programs 12.3-12.4 on a
    binary outcome): logistic propensity, stabilized weights, IP-weighted
    risk in each arm, return (RR, risk1, risk0)."""
    X = sm.add_constant(dd[covs].astype(float).values)
    A = dd[treat].astype(float).values
    Y = dd[y].astype(float).values
    den = sm.Logit(A, X).fit(disp=0)
    p_den = den.predict(X)
    p_num = A.mean()  # stabilizing numerator P(A=1)
    sw = np.where(A == 1, p_num / p_den, (1 - p_num) / (1 - p_den))
    r1 = np.sum(sw * A * Y) / np.sum(sw * A)
    r0 = np.sum(sw * (1 - A) * Y) / np.sum(sw * (1 - A))
    return r1 / r0, r1, r0


def _boot_rr_ci(dd, covs, y, treat, n_boot=1000, alpha=0.05, seed=42):
    rng = np.random.default_rng(seed)
    idx = np.arange(len(dd))
    ests = np.empty(n_boot)
    for b in range(n_boot):
        s = dd.iloc[rng.choice(idx, size=len(idx), replace=True)]
        try:
            ests[b] = _stabilized_ipw_rr(s, covs, y, treat)[0]
        except Exception:
            ests[b] = np.nan
    ests = ests[~np.isnan(ests)]
    lo, hi = np.percentile(ests, [100 * alpha / 2, 100 * (1 - alpha / 2)])
    return float(lo), float(hi)


def main() -> None:
    full = sp.datasets.nhefs()  # n=1629 -- death is complete
    full.to_csv(__import__("_common").DATA_DIR / f"{MODULE}.csv", index=False)
    n_full = len(full)

    dd_full, covs = book_design(full)
    A = dd_full["qsmk"].astype(float).values
    Y = dd_full["death"].astype(float).values

    # ---- PRIMARY: mortality risk ratios + E-values --------------------
    # Crude (unadjusted) RR -- CI excludes 1; the interesting E-value.
    risk1_c, risk0_c = Y[A == 1].mean(), Y[A == 0].mean()
    rr_crude = risk1_c / risk0_c
    # Wald CI on log-RR (closed form -> identical in R).
    n1, n0 = (A == 1).sum(), (A == 0).sum()
    se_log = math.sqrt((1 - risk1_c) / (risk1_c * n1) + (1 - risk0_c) / (risk0_c * n0))
    z = sp_norm_ppf(1 - 0.05 / 2)
    lo_c = rr_crude * math.exp(-z * se_log)
    hi_c = rr_crude * math.exp(z * se_log)

    # IP-weighted (adjusted) RR -- bootstrap CI (straddles 1 here).
    rr_adj, r1_adj, r0_adj = _stabilized_ipw_rr(dd_full, covs, "death", "qsmk")
    lo_a, hi_a = _boot_rr_ci(dd_full, covs, "death", "qsmk", n_boot=1000, seed=42)

    ev_crude = sp.evalue(estimate=rr_crude, ci=(lo_c, hi_c), measure="RR")
    ev_adj = sp.evalue(estimate=rr_adj, ci=(lo_a, hi_a), measure="RR")

    # Closed-form check for the crude point E-value.
    rr_used = rr_crude if rr_crude >= 1 else 1 / rr_crude
    closed_crude = rr_used + math.sqrt(rr_used * (rr_used - 1))

    # ---- SECONDARY: Ch12 weight effect (continuous) -------------------
    cc = sp.datasets.nhefs(complete_case=True)
    dd_cc, covs_cc = book_design(cc)
    wt_res = sp.ipw(
        dd_cc,
        y="wt82_71",
        treat="qsmk",
        covariates=covs_cc,
        estimand="ATE",
        seed=42,
        n_bootstrap=500,
    )
    effect = float(wt_res.estimate)  # ~3.44 kg (book 3.4)
    sd_out = float(cc["wt82_71"].std(ddof=1))
    d_smd = effect / sd_out
    # CI limits standardized too (closest-to-null limit drives evalue_ci).
    lo_d = float(wt_res.ci[0]) / sd_out
    hi_d = float(wt_res.ci[1]) / sd_out

    # Native StatsPAI SMD route (implements exp(0.91 d) internally).
    ev_smd = sp.evalue(estimate=d_smd, ci=(lo_d, hi_d), measure="SMD")
    # Manual VanderWeele-Ding RR_approx, passed as a plain RR, must agree.
    rr_approx = math.exp(VD_SMD_COEF * d_smd)
    rr_approx_lo = math.exp(VD_SMD_COEF * lo_d)
    rr_approx_hi = math.exp(VD_SMD_COEF * hi_d)
    ev_smd_manual = sp.evalue(
        estimate=rr_approx, ci=(rr_approx_lo, rr_approx_hi), measure="RR"
    )

    rows = [
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="evalue_crude_rr_point",
            estimate=float(ev_crude["evalue_estimate"]),
            se=None,
            n=int(n_full),
            published=closed_crude,
            citation="VanderWeele-Ding 2017 E = RR + sqrt(RR(RR-1)) (crude mortality RR)",
            extra={
                "rr": rr_crude,
                "rr_ci": [lo_c, hi_c],
                "evalue_ci": float(ev_crude["evalue_ci"]),
                "risk1": risk1_c,
                "risk0": risk0_c,
            },
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="evalue_ipw_rr_point",
            estimate=float(ev_adj["evalue_estimate"]),
            se=None,
            n=int(n_full),
            published=1.0,
            citation="VanderWeele-Ding 2017 E-value, IP-weighted mortality RR (~null after adjustment)",
            extra={
                "rr": rr_adj,
                "rr_ci": [lo_a, hi_a],
                "evalue_ci": float(ev_adj["evalue_ci"]),
                "risk1": float(r1_adj),
                "risk0": float(r0_adj),
            },
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="evalue_smd_point",
            estimate=float(ev_smd["evalue_estimate"]),
            se=None,
            n=int(len(cc)),
            published=float(ev_smd_manual["evalue_estimate"]),
            citation="VanderWeele-Ding SMD approx exp(0.91*d) on Ch12 weight effect",
            extra={
                "effect_kg": effect,
                "sd_outcome": sd_out,
                "d": d_smd,
                "rr_approx": float(ev_smd["rr_estimate"]),
                "rr_approx_manual": rr_approx,
                "evalue_ci": float(ev_smd["evalue_ci"]),
                "evalue_manual_point": float(ev_smd_manual["evalue_estimate"]),
            },
        ),
    ]

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "data_source": "sp.datasets.nhefs (NHEFS / What If)",
            "n_full": int(n_full),
            "n_complete": int(len(cc)),
            "vd_smd_coef": VD_SMD_COEF,
            "closed_form": "E = RR + sqrt(RR*(RR-1))",
            "smd_native_eq_manual": math.isclose(
                ev_smd["evalue_estimate"],
                ev_smd_manual["evalue_estimate"],
                rel_tol=1e-9,
            ),
        },
    )

    print(
        f"[{MODULE}] crude RR={rr_crude:.4f} CI({lo_c:.3f},{hi_c:.3f}) "
        f"E={ev_crude['evalue_estimate']:.4f} (closed {closed_crude:.4f}) "
        f"E_ci={ev_crude['evalue_ci']:.4f} || "
        f"ipw RR={rr_adj:.4f} CI({lo_a:.3f},{hi_a:.3f}) "
        f"E={ev_adj['evalue_estimate']:.4f} || "
        f"SMD d={d_smd:.4f} RR_approx={rr_approx:.4f} "
        f"E={ev_smd['evalue_estimate']:.4f} (manual {ev_smd_manual['evalue_estimate']:.4f})"
    )


def sp_norm_ppf(q: float) -> float:
    """Inverse standard-normal CDF (avoid importing scipy explicitly)."""
    from statistics import NormalDist

    return NormalDist().inv_cdf(q)


if __name__ == "__main__":
    main()

File: tests/orig_parity/11_nhefs_evalue.R
# Original-data parity for E-value sensitivity analysis (VanderWeele &
# Ding, Ann Intern Med 2017) on the REAL NHEFS / *What If* data.
#
#   Source data : sp.datasets.nhefs() FULL (n=1629; death is complete),
#                 dumped by the Python side to data/11_nhefs_evalue.csv so
#                 both languages read identical bytes.
#   Gold engine : the EValue R package (evalues.RR / evalues.MD) plus a
#                 hand-rolled stabilized-IPW risk ratio (base R glm), the
#                 same procedure the book uses for binary outcomes.
#
# The E-value is the deterministic VanderWeele-Ding closed form
#   E = RR + sqrt(RR*(RR-1))   (applied to 1/RR when RR<1),
# so this R gold must match StatsPAI's sp.evalue to ~1e-10.
#
#   PRIMARY  : crude mortality RR (qsmk) and IP-weighted mortality RR.
#   SECONDARY: Ch12 weight effect via the SMD approx RR=exp(0.91*d).
# Writes results/11_nhefs_evalue_R.json.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
HERE <- .script_dir
DATA_DIR <- file.path(HERE, "data")
RESULTS_DIR <- file.path(HERE, "results")
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(EValue)
  library(jsonlite)
})

MODULE <- "11_nhefs_evalue"
VD_SMD_COEF <- 0.91
df <- read.csv(file.path(DATA_DIR, paste0(MODULE, ".csv")))
n_full <- nrow(df)

A <- df$qsmk
Y <- df$death

# ---- PRIMARY: crude mortality risk ratio -----------------------------
risk1 <- mean(Y[A == 1]); risk0 <- mean(Y[A == 0])
rr_crude <- risk1 / risk0
n1 <- sum(A == 1); n0 <- sum(A == 0)
se_log <- sqrt((1 - risk1) / (risk1 * n1) + (1 - risk0) / (risk0 * n0))
z <- qnorm(0.975)
lo_c <- rr_crude * exp(-z * se_log)
hi_c <- rr_crude * exp(z * se_log)

ev_crude <- evalues.RR(est = rr_crude, lo = lo_c, hi = hi_c)
# evalues.RR returns a matrix: row "E-values", cols point/lower/upper.
ev_crude_point <- ev_crude["E-values", "point"]
ev_crude_ci    <- ev_crude["E-values", "lower"]  # CI limit closest to null

# Closed-form anchor.
rr_used <- if (rr_crude >= 1) rr_crude else 1 / rr_crude
closed_crude <- rr_used + sqrt(rr_used * (rr_used - 1))

# ---- PRIMARY: IP-weighted (stabilized) mortality risk ratio ----------
conf <- paste0(
  "factor(sex)+factor(race)+factor(education)+factor(exercise)+",
  "factor(active)+age+I(age^2)+smokeintensity+I(smokeintensity^2)+",
  "smokeyrs+I(smokeyrs^2)+wt71+I(wt71^2)"
)
ipw_rr <- function(d) {
  den <- glm(as.formula(paste("qsmk ~", conf)), data = d, family = binomial())
  pd  <- predict(den, type = "response")
  pn  <- mean(d$qsmk)
  sw  <- ifelse(d$qsmk == 1, pn / pd, (1 - pn) / (1 - pd))
  r1  <- sum(sw * d$qsmk * d$death) / sum(sw * d$qsmk)
  r0  <- sum(sw * (1 - d$qsmk) * d$death) / sum(sw * (1 - d$qsmk))
  r1 / r0
}
rr_adj <- ipw_rr(df)
set.seed(42)
B <- 1000
boot <- replicate(B, {
  idx <- sample.int(n_full, n_full, replace = TRUE)
  tryCatch(ipw_rr(df[idx, , drop = FALSE]), error = function(e) NA_real_)
})
boot <- boot[is.finite(boot)]
lo_a <- as.numeric(quantile(boot, 0.025))
hi_a <- as.numeric(quantile(boot, 0.975))
ev_adj <- evalues.RR(est = rr_adj, lo = lo_a, hi = hi_a)
ev_adj_point <- ev_adj["E-values", "point"]
ev_adj_ci    <- ev_adj["E-values", "upper"]  # RR<1 -> upper limit nearest null

# ---- SECONDARY: Ch12 weight effect via SMD approximation -------------
# Use the same complete-case weight effect StatsPAI reports (3.4-3.44 kg).
cc <- df[!is.na(df$wt82_71), ]
den <- glm(as.formula(paste("qsmk ~", conf)), data = cc, family = binomial())
pd  <- predict(den, type = "response")
pn  <- mean(cc$qsmk)
sw  <- ifelse(cc$qsmk == 1, pn / pd, (1 - pn) / (1 - pd))
msm <- lm(wt82_71 ~ qsmk, data = cc, weights = sw)
effect <- unname(coef(msm)["qsmk"])
sd_out <- sd(cc$wt82_71)
d_smd  <- effect / sd_out
rr_approx <- exp(VD_SMD_COEF * d_smd)
# EValue's evalues.MD standardizes MD internally as exp(0.91*d); here we
# already standardized, so feed evalues.RR(exp(0.91 d)) for a like-for-like
# check, AND evalues.MD on the raw (already-standardized) effect with se=0.
ev_smd <- evalues.RR(est = rr_approx)
ev_smd_point <- ev_smd["E-values", "point"]
ev_md  <- evalues.MD(est = d_smd, se = 0)        # exp(0.91*d) internally
ev_md_point <- ev_md["E-values", "point"]

rows <- list(
  list(module = MODULE, side = "R", statistic = "evalue_crude_rr_point",
       estimate = ev_crude_point, se = NULL, n = n_full,
       published = closed_crude,
       citation = "EValue::evalues.RR on crude mortality RR; closed form E=RR+sqrt(RR(RR-1))",
       extra = list(rr = rr_crude, rr_ci = c(lo_c, hi_c),
                    evalue_ci = ev_crude_ci, risk1 = risk1, risk0 = risk0)),
  list(module = MODULE, side = "R", statistic = "evalue_ipw_rr_point",
       estimate = ev_adj_point, se = NULL, n = n_full, published = 1.0,
       citation = "EValue::evalues.RR on IP-weighted mortality RR (~null after adjustment)",
       extra = list(rr = rr_adj, rr_ci = c(lo_a, hi_a),
                    evalue_ci = ev_adj_ci)),
  list(module = MODULE, side = "R", statistic = "evalue_smd_point",
       estimate = ev_smd_point, se = NULL, n = nrow(cc),
       published = ev_md_point,
       citation = "EValue::evalues.RR(exp(0.91 d)) vs evalues.MD on Ch12 weight effect",
       extra = list(effect_kg = effect, sd_outcome = sd_out, d = d_smd,
                    rr_approx = rr_approx,
                    evalue_md_point = ev_md_point))
)
payload <- list(
  module = MODULE, side = "R", rows = rows,
  extra = list(engine = "EValue::evalues.RR/MD + base-R stabilized IPW",
               n_full = n_full, n_complete = nrow(cc),
               vd_smd_coef = VD_SMD_COEF,
               closed_form = "E = RR + sqrt(RR*(RR-1))",
               smd_rr_eq_md = isTRUE(all.equal(ev_smd_point, ev_md_point)))
)
writeLines(toJSON(payload, auto_unbox = TRUE, null = "null", digits = 10),
           file.path(RESULTS_DIR, paste0(MODULE, "_R.json")))
cat(sprintf(paste0("[%s] crude RR=%.4f CI(%.3f,%.3f) E=%.4f (closed %.4f) ",
                   "E_ci=%.4f || ipw RR=%.4f CI(%.3f,%.3f) E=%.4f || ",
                   "SMD d=%.4f RR_approx=%.4f E=%.4f (MD %.4f)\n"),
            MODULE, rr_crude, lo_c, hi_c, ev_crude_point, closed_crude,
            ev_crude_ci, rr_adj, lo_a, hi_a, ev_adj_point,
            d_smd, rr_approx, ev_smd_point, ev_md_point))
